Revise security guide (#1)
nupuruttarwar/networking-recipe: docs/guides/installing-tls-certificates.md

# Generating and installing TLS certificates

> - [1. Overview](#1-overview)

Author: Is this a good known practice about level and level 3 cases for markdown files?

- [2. Generate and install TLS certificates](#2-generate-and-install-tls-certifificates)
   - [2.1 Generate certificates](#21-generate-certificates)
   - [2.2 Install certificates](#22-install-certificates)
       - [2.2.1 Install at default location](#221-install-at-default-location)
       - [2.2.2 Install at non-default location](#222-install-at--non--default-location)

## 1. Overview

This document provides information about generating and installing TLS certificates
for running infrap4d in secure mode

> ## 2. Generate and install TLS certificates

Author: Is this a good known practice about level and level 3 cases for markdown files?

Note: Here target name can be `dpdk` or `es2k`

### 2.1 Generate certificates

> - Review the files `ca.conf` and `grpc-client.conf` available under

Author: Okay, will get rid of the bullets. User can generate the certs on any system and then copy it to ACC and (server) and client can be ACC or Host.

[Response to 3] Code block was indented because it was under bullet and if you flush it to left, it doesn't look good when you view the file

[Response to 4] IP is the IP address of server

[Response to 5] Which non-example text are you referring to here?

[Response to 6] Okay, will remove the note. 

[Response to 7] copy_config_files install these generate_certs.sh under /usr/share/stratum/es2k

[Response to 8] Sabeel had developed it this, didn't want changing too many things at once.

I think we're missing some procedural context. What platform is the user running on when they invoke generate-certs.sh? The development system? The IMC? The ACC? If it's the IMC or ACC, how did they get there?
Response: Ideally they can run generate_certs.sh anywhere. Since, we provide generate_certs.sh, they can run it on Host or ACC and it will be available under /usr/share/stratum/<target_name>.

generate_cert.sh is just an example on how to generate certificates. They can use other known methods for generating TLS certificates

> - Review the files `ca.conf` and `grpc-client.conf` available under

Reviewer: > [Response to 5] Which non-example text are you referring to here?

The word "or".

Conditional variations in command-line examples are infuriatingly confusing. There are a number of these in the P4 SDE build instructions. It's easy to get confused following them. I really dislike them. Better to have a series of complete examples with differing subheadings and explanatory text.

> - Review the files `ca.conf` and `grpc-client.conf` available under

Reviewer: > [Response to 3] Code block was indented because it was under bullet and if you flush it to left, it doesn't look good when you view the file

It doesn't look good with the bullet and indented text either. The style I'm promoting is similar to the "modified block format" (for letters) that I learned in my high school typing class.

# Generating and installing TLS certificates

- [1. Overview](#1-overview)
- [2. Generating TLS certificates](#2-generating-tls-certifificates)
- [3. Installing TLS certificates](#3-installing-tls-certificates)
    - [3.1 Install at default location](#31-install-at-default-location)
    - [3.2 Install at non-default location](#32-install-at--non--default-location)

## 1. Overview

This document provides information about generating and installing TLS certificates
for running infrap4d in secure mode

## 2. Generating TLS certificates
Note: Here target name can be `dpdk` or `es2k`

- Review the files `ca.conf` and `grpc-client.conf` available under
`/usr/share/stratum/<target_name>` to verify that the configuration settings are
as desired
- Run the `generate-certs.sh` available under `/usr/share/stratum/<target_name>`

```bash
	cd /usr/share/stratum/<target_name>/

	COMMON_NAME=<IP> ./generate-certs.sh
	or
	COMMON_NAME=localhost ./generate-certs.sh
```

Note: System relies on mTLS (mutual TLS) for authentication.

## 3. Installing TLS certificates

`infrap4d` will check for server certificates in the default location
`/usr/share/stratum/certs/`

### 3.1 Install at default location

> - Copy the generated `ca.crt`, `stratum.crt`, and `stratum.key` to the server

Reviewer: This is an ordinary paragraph, not a bullet.

running infrap4d:

> 

Reviewer: One blank line between paragraphs.

```bash

> mkdir -p /usr/share/stratum/certs

Reviewer: I don't understand the context.

Why copy the certificates from `/usr/share/stratum/<target_name>/certs` to `/usr/share/stratum/certs` on the same platform? Why not simply generate them there? Better still, why not have the script and configuration files in `/usr/share/stratum/certs` generate the key and certificate files in place?

If you're copying them to a _different_ platform, there's nothing in the instructions that make that celar.

Author: This is how the code and scripts were written initially.

generate-certs.sh will generate certs folder having all certs. This will be copied to /usr/share/stratum/certs which is the default location.

I have said "Copy the generated `ca.crt`, `stratum.crt`, and `stratum.key` to the server running infrap4d". In our case, it can he host or ACC. Same applies for client too.

cp /usr/share/stratum/<target_name>/certs/stratum.crt /usr/share/stratum/certs/
cp /usr/share/stratum/<target_name>/certs/ca.crt /usr/share/stratum/certs/
cp /usr/share/stratum/<target_name>/certs/stratum.key /usr/share/stratum/certs/
```

> - Copy the generated `ca.crt`, `client.crt`, and `client.key` to the

Reviewer: This is an ordinary paragraph, not a bullet.

client machine

> 

Reviewer: One blank line before opening fence.

```bash
mkdir -p /usr/share/stratum/certs

> cp /usr/share/stratum/<target_name>/certs/ca.crt /usr/share/stratum/certs/

Reviewer: This illustrates copying files within same file system. It does not illustrate copying files from one system to another.

Author: Should I cover a case separately where server is on ACC and client is on Host or just say copy and not provide examples as it will complicate it?

cp /usr/share/stratum/<target_name>/certs/client.crt /usr/share/stratum/certs/
cp /usr/share/stratum/<target_name>/certs/client.key /usr/share/stratum/certs/
```

### 3.2 Install at non-default location

> If alternate location is desired for server certificates, copy the certifactes

Reviewer: ```suggestion
If you would like to use a different location for the server certificates, copy the certificates
```

> at desired location and the location must be specified with following options

Reviewer: ```suggestion
to the location and specify the following options
```

> when running infrap4d.

Reviewer: ```suggestion
```

> when running infrap4d.

Reviewer: ```suggestion
on the `infrap4d` command line.
```

> ```text

Reviewer: I'd suggest using a table instead of a text block.

| Option | Description |
| ------ | ----------- |
| `-ca_cert_file=PATH` | CA certificate file |
| `-server_cert_file=PATH` | Server certificate file |
| `-server_key_file=PATH` | Server private key file |

-ca_cert_file=[CA certificate file]
-server_cert_file=[Server certificate file]
-server_key_file=[Server private key file]
```

For example:

> ```bash

Reviewer: The examples do not make it clear what you're doing.

Are you moving files between systems? Between directories on a shared disk?

# On server, copy the certs at non-default location /tmp/certs
mkdir -p /tmp/certs
cp /usr/share/stratum/<target_name>/certs/stratum.crt /tmp/certs
cp /usr/share/stratum/<target_name>/certs/ca.crt /tmp/certs
cp /usr/share/stratum/<target_name>/certs/stratum.key /tmp/certs

# On client, copy the certs at non-default location /tmp/certs
mkdir -p /tmp/certs
cp /usr/share/stratum/<target_name>/certs/ca.crt /tmp/certs
cp /usr/share/stratum/<target_name>/certs/client.crt /tmp/certs
cp /usr/share/stratum/<target_name>/certs/client.key /tmp/certs

# On server, start infrap4d
$IPDK_RECIPE/install/sbin/infrap4d \
    -ca_cert_file=/tmp/certs/ca.crt \
    -server_cert_file=/tmp/certs/stratum.crt \
    -server_key_file=/tmp/certs/stratum.key
```

> Note: Alternate location for client certificates is not supported. For more details about available

Reviewer: Alternate? An alternative to what?

Author: to a default location. Should I say non-default location instead?

Reviewer: As written, this is a sort of double negative ("not in the default location is not supported"). And "default" suggests that there  are choices; you're basically saying that there aren't any. Maybe just flip it around, and say that "the client certificates must be installed in (whatever the standard location is)."
